### PR TITLE
Add global alpha and colorkey info to Surf repr

### PR DESCRIPTION
**Pull request by SSS-Says-Snek**
Implements #2014

### CODE REVIEW

**Comment by SSS-Says-Snek** [comment]
I think I need to cast the self ptr, I'll try when I get back

**Review by yunline — APPROVED** [review]
LGTM 👍

**Comment by Starbuck5** [comment]
It seems like it should still have the parentheses, even though it's not a real constructor.

`<Surface 20x20x32>` -> `<Surface(20x20x32)>`

File: src_c/surface.c [hunk 1]
```diff
@@ -499,8 +499,23 @@ surface_str(PyObject *self)
...
-                                surf->format->BitsPerPixel);
+    PyObject *colorkey = surf_get_colorkey((pgSurfaceObject *)self, NULL);
+    PyObject *global_alpha = surf_get_alpha((pgSurfaceObject *)self, NULL);
+
+    // 46 is max len of format str, plus null terminator
+    char format[47] = "<Surface %dx%dx%d";
+    if (PyObject_IsTrue(colorkey))
+        strcat(format, " colorkey=%S");
+    if (PyObject_IsTrue(global_alpha))
+        strcat(format, " global_alpha=%S");
```
Comment (itzpr3d4t0r) [inline src_c/surface.c:510]: I think there should be commas in between the different "attributes" so like:
```suggestion
        strcat(format, ", colorkey=%S");
    if (PyObject_IsTrue(global_alpha))
        strcat(format, ", global_alpha=%S");
```

Comment (SSS-Says-Snek) [reply src_c/surface.c:510]: Sure, I was mainly following from the issue

**Comment by SSS-Says-Snek** [comment]
I'll add those things when I get home

File: src_c/surface.c [hunk 2]
```diff
@@ -499,8 +499,23 @@ surface_str(PyObject *self)
...
-                                surf->format->BitsPerPixel);
+    PyObject *colorkey = surf_get_colorkey((pgSurfaceObject *)self, NULL);
+    PyObject *global_alpha = surf_get_alpha((pgSurfaceObject *)self, NULL);
+
+    // 49 is max len of format str, plus null terminator
+    char format[50] = "<Surface(%dx%dx%d";
+    if (PyObject_IsTrue(colorkey))
+        strcat(format, ", colorkey=%S");
+    if (PyObject_IsTrue(global_alpha))
+        strcat(format, ", global_alpha=%S");
```
[the remark this answers is not in the document]
Comment (SSS-Says-Snek) [reply src_c/surface.c:510]: To me, the parentheses surrounding each part of the repr as suggested in the upstream issue is redundant now that commas separate them

File: src_c/surface.c [hunk 3]
```diff
@@ -499,8 +499,23 @@ surface_str(PyObject *self)
         return PyUnicode_FromString("<Surface(Dead Display)>");
     }
 
-    return PyUnicode_FromFormat("<Surface(%dx%dx%d SW)>", surf->w, surf->h,
-                                surf->format->BitsPerPixel);
+    PyObject *colorkey = surf_get_colorkey((pgSurfaceObject *)self, NULL);
+    PyObject *global_alpha = surf_get_alpha((pgSurfaceObject *)self, NULL);
```
[the remark this answers is not in the document]
Comment (SSS-Says-Snek) [reply src_c/surface.c:503]: Yeah, maybe we can directly find the alpha

File: src_c/surface.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (itzpr3d4t0r) [reply src_c/surface.c:510]: Sry i meant curly braces:
```suggestion
    if (PyObject_IsTrue(colorkey)) {
        strcat(format, ", colorkey=%S");
        }
    if (PyObject_IsTrue(global_alpha)) {
        strcat(format, ", global_alpha=%S");
        }
```

File: src_c/surface.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (SSS-Says-Snek) [reply src_c/surface.c:503]: Now that I think about it, we can tolerate a small perf hit, as people don't really print surfaces that often, and even so, it's probably just for debug

File: src_c/surface.c [hunk 4]
```diff
@@ -499,8 +499,25 @@ surface_str(PyObject *self)
...
```
Comment (Starbuck5) [inline src_c/surface.c:503]: These are both leaking memory.

Comment (MyreMylar) [reply src_c/surface.c:503]: you should be able to fix this with:
```
Py_DECREF(colorkey);
Py_DECREF(global_alpha);
```

Once you are done with the objects, perhaps on line 514.

Comment (itzpr3d4t0r) [reply src_c/surface.c:503]: No he has to put the `PyUnicode_FromFormat` at line 518 in a separate variable before the return, DECREF them and the nreturn the variable.

Comment (SSS-Says-Snek) [reply src_c/surface.c:503]: Oops, IG I thought that PyUnicode_FromFormat eats the references

**Review by itzpr3d4t0r — APPROVED** [review]
LGTM

File: src_c/surface.c [hunk 5]
```diff
@@ -499,8 +499,30 @@ surface_str(PyObject *self)
...
```
Comment (Starbuck5) [inline src_c/surface.c:503]: Sorry to not spot this earlier, but I believe these should forward errors

I.e.

```c
PyObject *colorkey = surf_get_colorkey((pgSurfaceObject *)self, NULL);
if (colorkey == NULL) {
    return NULL;
}
PyObject *global_alpha = surf_get_alpha((pgSurfaceObject *)self, NULL);
if (global_alpha == NULL) {
    Py_DECREF(colorkey);
    return NULL;
}
```

**Review by Starbuck5 — APPROVED** [review]

**Pull request merged**